### PR TITLE
Support videos in Android CameraRoll.getPhotos

### PR DESCRIPTION
**Pull request by kevinresol**
Support getting videos for `CameraRoll.getPhotos()` on Android. Before this PR, `assetType` is ignored in Android.

### CODE REVIEW

File: ReactAndroid/src/main/java/com/facebook/react/modules/camera/CameraRollManager.java [hunk 1]
```diff
@@ -279,6 +286,19 @@ protected void doInBackgroundGuarded(Void... params) {
         selection.append(" AND " + SELECTION_BUCKET);
         selectionArgs.add(mGroupName);
       }
+      if (!TextUtils.isEmpty(mAssetType)) {
+        if(mAssetType.equals("Videos")) {
```
Comment (hramos) [inline ReactAndroid/src/main/java/com/facebook/react/modules/camera/CameraRollManager.java]: Can you compare against a constant here instead of the string, "Videos"?

File: ReactAndroid/src/main/java/com/facebook/react/modules/camera/CameraRollManager.java [hunk 2]
```diff
@@ -226,6 +228,7 @@ public void getPhotos(final ReadableMap params, final Promise promise) {
     int first = params.getInt("first");
     String after = params.hasKey("after") ? params.getString("after") : null;
     String groupName = params.hasKey("groupName") ? params.getString("groupName") : null;
+    String assetType = params.hasKey("assetType") ? params.getString("assetType") : "Photos";
```
Comment (hramos) [inline ReactAndroid/src/main/java/com/facebook/react/modules/camera/CameraRollManager.java]: Can you use a constant here instead of "Photos"?

File: ReactAndroid/src/main/java/com/facebook/react/modules/camera/CameraRollManager.java [hunk 3]
```diff
@@ -294,7 +314,7 @@ protected void doInBackgroundGuarded(Void... params) {
       // an SQLite DB and forwards parameters to it without doing any parsing / validation.
       try {
         Cursor photos = resolver.query(
-            Images.Media.EXTERNAL_CONTENT_URI,
+            MediaStore.Files.getContentUri("external"),
```
Comment (hramos) [inline ReactAndroid/src/main/java/com/facebook/react/modules/camera/CameraRollManager.java]: Can you explain why this change is needed?

Comment (kevinresol) [inline ReactAndroid/src/main/java/com/facebook/react/modules/camera/CameraRollManager.java]: This is because querying against `Images.Media.EXTERNAL_CONTENT_URI` will only gives you images. So we need to query against the "whole external content uri" and use `MediaStore.Files.FileColumns.MEDIA_TYPE` to filter out what we need, either photos or videos or both.

File: ReactAndroid/src/main/java/com/facebook/react/modules/camera/CameraRollManager.java (hunk 2 above)
Comment (kevinresol) [inline ReactAndroid/src/main/java/com/facebook/react/modules/camera/CameraRollManager.java]: constantsToExport

Comment (hramos) [inline ReactAndroid/src/main/java/com/facebook/react/modules/camera/CameraRollManager.java]: I meant using a constant in Java, but after taking another look at this file, it seems like that would be against convention. Disregard.

**Comment by kevinresol** [comment]
> Approved. It looks like tests are not passing at this time, however.

I am not very familiar with RN's test suite. Is there something I should do here?

File: ReactAndroid/src/main/java/com/facebook/react/modules/camera/CameraRollManager.java [hunk 4]
```diff
@@ -347,12 +367,13 @@ private static void putEdges(
     int heightIndex = IS_JELLY_BEAN_OR_LATER ? photos.getColumnIndex(Images.Media.HEIGHT) : -1;
     int longitudeIndex = photos.getColumnIndex(Images.Media.LONGITUDE);
     int latitudeIndex = photos.getColumnIndex(Images.Media.LATITUDE);
+    int dataIndex = photos.getColumnIndex(MediaStore.MediaColumns.DATA);
```
Comment (mkonicek) [inline ReactAndroid/src/main/java/com/facebook/react/modules/camera/CameraRollManager.java]: This variable should not be called `photos` anymore since it's either photos or vides, correct? How about `media`?

Comment (mkonicek) [inline ReactAndroid/src/main/java/com/facebook/react/modules/camera/CameraRollManager.java]: Why is `dataIndex` needed when it wasn't needed previously. Please add a comment above this line explaining what `dataIndex` is.

File: ReactAndroid/src/main/java/com/facebook/react/modules/camera/CameraRollManager.java (hunk 1 above)
Comment (mkonicek) [inline ReactAndroid/src/main/java/com/facebook/react/modules/camera/CameraRollManager.java]: nit: Space after `if`

File: ReactAndroid/src/main/java/com/facebook/react/modules/camera/CameraRollManager.java [hunk 5]
```diff
@@ -279,6 +286,19 @@ protected void doInBackgroundGuarded(Void... params) {
         selection.append(" AND " + SELECTION_BUCKET);
         selectionArgs.add(mGroupName);
       }
+      if (!TextUtils.isEmpty(mAssetType)) {
+        if(mAssetType.equals("Videos")) {
+          selection.append(" AND " + MediaStore.Files.FileColumns.MEDIA_TYPE + " = "
+            + MediaStore.Files.FileColumns.MEDIA_TYPE_VIDEO);
+        } else if(mAssetType.equals("Photos")) {
```
Comment (mkonicek) [inline ReactAndroid/src/main/java/com/facebook/react/modules/camera/CameraRollManager.java]: space after `if`

File: ReactAndroid/src/main/java/com/facebook/react/modules/camera/CameraRollManager.java [hunk 6]
```diff
@@ -248,6 +252,7 @@ public void getPhotos(final ReadableMap params, final Promise promise) {
     private final int mFirst;
     private final @Nullable String mAfter;
     private final @Nullable String mGroupName;
+    private final @Nullable String mAssetType;
```
Comment (mkonicek) [inline ReactAndroid/src/main/java/com/facebook/react/modules/camera/CameraRollManager.java]: Remove `@Nullable` from here because your code below expects it's always "Photos" or "Videos".

File: ReactAndroid/src/main/java/com/facebook/react/modules/camera/CameraRollManager.java [hunk 7]
```diff
@@ -256,13 +261,15 @@ private GetPhotosTask(
...
         @Nullable String groupName,
+        @Nullable String assetType,
         @Nullable ReadableArray mimeTypes,
         Promise promise) {
       super(context);
       mContext = context;
       mFirst = first;
       mAfter = after;
       mGroupName = groupName;
+      mAssetType = assetType;
```
Comment (mkonicek) [inline ReactAndroid/src/main/java/com/facebook/react/modules/camera/CameraRollManager.java]: Throw a NPE if `null` is passed.

File: ReactAndroid/src/main/java/com/facebook/react/modules/camera/CameraRollManager.java [hunk 8]
```diff
@@ -256,13 +261,15 @@ private GetPhotosTask(
         int first,
         @Nullable String after,
         @Nullable String groupName,
+        @Nullable String assetType,
```
Comment (mkonicek) [inline ReactAndroid/src/main/java/com/facebook/react/modules/camera/CameraRollManager.java]: Remove `@Nullable`.

File: ReactAndroid/src/main/java/com/facebook/react/modules/camera/CameraRollManager.java [hunk 9]
```diff
@@ -279,6 +286,19 @@ protected void doInBackgroundGuarded(Void... params) {
         selection.append(" AND " + SELECTION_BUCKET);
         selectionArgs.add(mGroupName);
       }
+      if (!TextUtils.isEmpty(mAssetType)) {
```
Comment (mkonicek) [inline ReactAndroid/src/main/java/com/facebook/react/modules/camera/CameraRollManager.java]: This can't be empty - correct? Otherwise the code below won't filter photos or videos. What would the method return then?

File: ReactAndroid/src/main/java/com/facebook/react/modules/camera/CameraRollManager.java (hunk 3 above)
Comment (mkonicek) [inline ReactAndroid/src/main/java/com/facebook/react/modules/camera/CameraRollManager.java]: Is there an Android constant for this? Is this the same as `Environment.getExternalStorageDirectory()`?

http://stackoverflow.com/questions/17215466/environment-getexternalstoragedirectory-getabsolutepath-gives-a-different-pa

Prefer using a constant.

File: ReactAndroid/src/main/java/com/facebook/react/modules/camera/CameraRollManager.java [hunk 10]
```diff
@@ -386,11 +407,10 @@ private static boolean putImageInfo(
...
       int idIndex,
       int widthIndex,
-      int heightIndex) {
+      int heightIndex,
+      int dataIndex) {
     WritableMap image = new WritableNativeMap();
-    Uri photoUri = Uri.withAppendedPath(
-        Images.Media.EXTERNAL_CONTENT_URI,
-        photos.getString(idIndex));
+    Uri photoUri = Uri.parse("file://" + photos.getString(dataIndex));
```
Comment (mkonicek) [inline ReactAndroid/src/main/java/com/facebook/react/modules/camera/CameraRollManager.java]: Why get rid of the `EXTERNAL_CONTENT_URI` here and not use ` MediaStore.Files.getContentUri("external")` like above?

Comment (kevinresol) [inline ReactAndroid/src/main/java/com/facebook/react/modules/camera/CameraRollManager.java]: Because it simply won't work to append the id directly to the external content uri.

File: ReactAndroid/src/main/java/com/facebook/react/modules/camera/CameraRollManager.java (hunk 4 above)
Comment (kevinresol) [inline ReactAndroid/src/main/java/com/facebook/react/modules/camera/CameraRollManager.java]: I tried to limit the amount of changes to functional parts only. There are quite a few things that can be improved but I am just afraid to expand the scope of this PR too much.

dataIndex is needed for extracting the uri of the file. I found this the easiest way to reliably obtain the file uri.

**Comment by gevorg94** [comment]
is this issue already resolved?

**Comment by Ehesp** [comment]
@kevinresol @mkonicek Would it be possible to make headway on this? As far as I can see videos still don't show from the camera roll.

**Comment by kevinresol** [comment]
Sorry but I can't maintain this anymore. If anyone is interested feel free to fork from my branch and create another PR.

**Comment by kesha-antonov** [comment]
Hey all.
I've created new PR: https://github.com/facebook/react-native/pull/13715

Please vote up to get it merged ASAP!

File: ReactAndroid/src/main/java/com/facebook/react/modules/camera/CameraRollManager.java (hunk 1 above)
Comment (ManikSmartdata) [reply ReactAndroid/src/main/java/com/facebook/react/modules/camera/CameraRollManager.java]: Do i need to to upgrade my RN to 0.47.2. because below version doesn't have any video support

Comment (hramos) [reply ReactAndroid/src/main/java/com/facebook/react/modules/camera/CameraRollManager.java]: This PR was not merged.

Comment (ManikSmartdata) [reply ReactAndroid/src/main/java/com/facebook/react/modules/camera/CameraRollManager.java]: I am using RN version 0.45.0. Video support is not there in 0.45.0. What should I do. Please let me know
